src/constructors.jl

```diff
@@ -58,130 +19,45 @@ function Arb(str::AbstractString; prec::Integer = DEFAULT_PRECISION[])
 end
 
-function Acb(z::Complex{Arb}; prec::Integer = max(precision(real(z)), precision(imag(z))))
+function Acb(re::AbstractString, im::AbstractString; prec::Integer = DEFAULT_PRECISION[])
     res = Acb(prec = prec)
-    set!(res, real(z), imag(z))
+    flag = set!(realref(res), re)
+    iszero(flag) || throw(ArgumentError("could not parse $str as an Arb"))
+    flag = set!(imagref(res), im)
+    iszero(flag) || throw(ArgumentError("could not parse $str as an Arb"))
     return res
 end
 
-Acb(re::Real, im::Real; prec::Integer = max(_precision(re), _precision(im))) =
-    Acb(Arb(re, prec = prec), Arb(im, prec = prec))
-
-Acb(z::Complex; prec::Integer = max(_precision(real(z)), _precision(imag(z)))) =
-    Acb(real(z), imag(z), prec = prec)
-
 Base.zero(::Union{Mag,Type{Mag}}) = Mag(UInt64(0))
 Base.one(::Union{Mag,Type{Mag}}) = Mag(UInt64(1))
 Base.zero(x::T) where {T<:Union{Arf,Arb,Acb}} = T(0, prec = precision(x))
 Base.one(x::T) where {T<:Union{Arf,Arb,Acb}} = T(1, prec = precision(x))
-Base.zero(x::AcbRef) = Acb(0, prec = precision(x))
-Base.one(x::AcbRef) = Acb(1, prec = precision(x))
-Base.zero(x::ArbRef) = Arb(0, prec = precision(x))
-Base.one(x::ArbRef) = Arb(1, prec = precision(x))
+
 # Define these since the base implementation would create `n` copies of the same element
 # I.e. only allocating **one** Arf/Arb/Acb.
-Base.zeros(x::T, n::Integer) where {T<:Union{Arf,Arb,Acb}} = [zero(x) for _ = 1:n]
-Base.ones(x::T, n::Integer) where {T<:Union{Arf,Arb,Acb}} = [one(x) for _ = 1:n]
-Base.zeros(x::Type{T}, n::Integer) where {T<:Union{Arf,Arb,Acb}} = [zero(T) for _ = 1:n]
-Base.ones(x::Type{T}, n::Integer) where {T<:Union{Arf,Arb,Acb}} = [one(T) for _ = 1:n]
-
-set!(z::Union{Acb,AcbRef,Ptr{acb_struct},acb_struct}, x::Complex{<:Base.GMP.CdoubleMax}) =
-    set!(z, real(x), imag(x))
-
-function set!(z::ArbLike, x::Rational)
-    z[] = numerator(x)
-    div!(z, z, denominator(x))
-end
-
-function set!(z::AcbLike, x::Union{Rational,Complex{<:Rational}})
-    set!(realref(z), real(x))
-    set!(imagref(z), imag(x))
-    z
-end
-
-# Irrationals
-function Mag(::Irrational{:π})
-    res = Mag()
-    const_pi!(res)
-    return res
-end
-
-for (irr, suffix) in ((:π, "pi"), (:ℯ, "e"), (:γ, "euler"))
-    jlf = Symbol("const_$suffix", "!")
-    IrrT = Irrational{irr}
-    @eval begin
-        function Arb(::$IrrT; prec::Integer = DEFAULT_PRECISION[])
-            res = Arb(prec = prec)
-            $jlf(res)
-            return res
-        end
-    end
-end
-
-function Acb(::Irrational{:π}; prec::Integer = DEFAULT_PRECISION[])
-    res = Acb(prec = prec)
-    const_pi!(res)
-    return res
-end
+Base.zeros(x::T, n::Integer) where {T<:Union{Mag,Arf,Arb,Acb}} = [zero(x) for _ = 1:n]
+Base.ones(x::T, n::Integer) where {T<:Union{Mag,Arf,Arb,Acb}} = [one(x) for _ = 1:n]
+Base.zeros(::Type{T}, n::Integer) where {T<:Union{Mag,Arf,Arb,Acb}} = [zero(T) for _ = 1:n]
+Base.ones(::Type{T}, n::Integer) where {T<:Union{Mag,Arf,Arb,Acb}} = [one(T) for _ = 1:n]
```

Comment:
This is a question: should we be returning ArbVectors/ArbMatrices here?

Reply:
I don't think so. It would make the interface a bit confusing since what should it return in case you want more dimensions?  Writing this I realise that `zeros(Arb, n1, n2)` actually doesn't work... Neither does `zeros(Arb, n1, n2, n3, ...)`. This we should fix.